keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply the
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to the match rules of admission webhooks, RBAC, priority & fairness and Audit)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  but with access to the data in
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to admission webhook's MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+failure policies, type safety, advanced matching rules, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+spec:
+  paramSource:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  denyReason: Invalid
+  validations:
+    - name: max-replicas
+      expression: "object.spec.replicas <= params.maxReplicas"
+      message: "object.spec.replicas must be no greater than {1}"
```

Comment:
This was my first choice (see the alternatives section at the bottom).

I could still be convinced to go another way here. The current down sides of using CEL directly seem to be:

- This example requires explicit cases, e.g. `"object.spec.replicas must be no greater than " + string(params.maxReplicas)`
- a plain string message had to be double quoted, e.g. `"'this is a simple message'`. Note that for CRD validation we only allow messages to be plain strings, so this creates an inconsistency. CRD validation would allow `message: "this is a simple message"` but admission would not.

The second one is more concerning for me than the first. We could offer both a `message` and `messageExpression` field (mutually exclusive)..

Reply:
That's right, I was thinking of a feature I'd used to treat fields as string literals or CEL. If parsed as valid CEL, I treated it as such, then compiled. Otherwise, I converted the string to a valid CEL string literal and turned that into the program. It's a bit of a hack, and I'm not sure it would work in this case, but it certainly makes the field easy to use. 

I'm also totally open to exposing a `string.format(string, list(dyn))` or `string.format(string, map(string, dyn))` function. I'd probably try to use `fmt.Sprintf` style strings since the format is well-implemented and relatively consistent across C++, Java, and Go. If it can convert to a string, then the `string(<value>)` will be applied with special handling for `list` and `map` values.

Reply:
SGTM, I'll switch over to `message` and `messageExpression` and make a note about `string.format()` (could be receiver fn? `"%d must be less than %d".format([value, max])`)